planner/core/rule_predicate_push_down.go

```diff
@@ -439,6 +439,95 @@ func (p *LogicalUnionAll) PredicatePushDown(predicates []expression.Expression,
+			if !groupByColumns.Contains(col) {
+				ok = false
+				break
+			}
+		}
+		if ok {
+			newFunc := expression.ColumnSubstitute(cond, la.Schema(), exprsOriginal)
+			condsToPush = append(condsToPush, newFunc)
+		} else {
+			ret = append(ret, cond)
+		}
+	default:
+		ret = append(ret, cond)
+	}
+	return condsToPush, ret
+}
+
+// pushDownPredicatesForAggregation split a CNF condition to two parts, can be pushed-down or can not be pushed-down below aggregation.
+// It would consider the CNF.
+// For example,
+// (a > 1 or avg(b) > 1) and (a < 3), and `avg(b) > 1` can't be pushed-down.
+// Then condsToPush: a < 3, ret: a > 1 or avg(b) > 1
+func (la *LogicalAggregation) pushDownCNFPredicatesForAggregation(cond expression.Expression, groupByColumns *expression.Schema, exprsOriginal []expression.Expression) ([]expression.Expression, []expression.Expression) {
+	var condsToPush []expression.Expression
+	var ret []expression.Expression
+	subCNFItem := expression.SplitCNFItems(cond)
+	if len(subCNFItem) == 1 {
+		return la.pushDownPredicatesForAggregation(subCNFItem[0], groupByColumns, exprsOriginal)
+	}
+	for _, item := range subCNFItem {
+		condsToPushForItem, retForItem := la.pushDownDNFPredicatesForAggregation(item, groupByColumns, exprsOriginal)
+		if len(condsToPushForItem) > 0 {
+			condsToPush = append(condsToPush, expression.ComposeDNFCondition(la.ctx, condsToPushForItem...))
+		}
+		if len(retForItem) > 0 {
+			ret = append(ret, expression.ComposeDNFCondition(la.ctx, retForItem...))
+		}
+	}
+	return condsToPush, ret
+}
+
+// pushDownDNFPredicatesForAggregation split a DNF condition to two parts, can be pushed-down or can not be pushed-down below aggregation.
+// It would consider the DNF.
+// For example,
+// (a > 1 and avg(b) > 1) or (a < 3), and `avg(b) > 1` can't be pushed-down.
+// Then condsToPush: (a < 3) and (a > 1), ret: (a > 1 and avg(b) > 1) or (a < 3)
+func (la *LogicalAggregation) pushDownDNFPredicatesForAggregation(cond expression.Expression, groupByColumns *expression.Schema, exprsOriginal []expression.Expression) ([]expression.Expression, []expression.Expression) {
+	var condsToPush []expression.Expression
+	var ret []expression.Expression
+	subDNFItem := expression.SplitDNFItems(cond)
+	if len(subDNFItem) == 1 {
+		return la.pushDownPredicatesForAggregation(subDNFItem[0], groupByColumns, exprsOriginal)
+	}
+	for _, item := range subDNFItem {
+		condsToPushForItem, retForItem := la.pushDownCNFPredicatesForAggregation(item, groupByColumns, exprsOriginal)
+		if len(condsToPushForItem) > 0 {
+			condsToPush = append(condsToPush, expression.ComposeCNFCondition(la.ctx, condsToPushForItem...))
+		} else {
+			return nil, []expression.Expression{cond}
+		}
+		if len(retForItem) > 0 {
+			ret = append(ret, expression.ComposeCNFCondition(la.ctx, retForItem...))
+		}
+	}
+	dnfPushDownCond := expression.ComposeDNFCondition(la.ctx, condsToPush...)
+	dnfPushDownCond = expression.FoldConstant(dnfPushDownCond)
+	if len(ret) == 0 {
```

Comment:
If there are no remained ones. We can directly push down the original one. Don't need to build one again.